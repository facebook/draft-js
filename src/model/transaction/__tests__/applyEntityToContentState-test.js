/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 */

'use strict';

jest.autoMockOff();

const Immutable = require('immutable');
const SelectionState = require('SelectionState');
const applyEntityToContentState = require('applyEntityToContentState');
const getSampleStateForTesting = require('getSampleStateForTesting');

describe('applyEntityToContentState', () => {
  const {
    contentState,
  } = getSampleStateForTesting();

  function checkForCharacterList(block) {
    expect(Immutable.List.isList(block.getCharacterList())).toBe(true);
  }

  function getEntities(block) {
    return block.getCharacterList().map(c => c.getEntity()).toJS();
  }

  describe('Apply entity within single block', () => {
    const target = contentState.getBlockMap().first();
    const targetSelection = new SelectionState({
      anchorKey: target.getKey(),
      anchorOffset: 0,
      focusKey: target.getKey(),
      focusOffset: target.getLength(),
    });

    function applyAndCheck(entityKey) {
      const withNewEntity = applyEntityToContentState(
        contentState,
        targetSelection,
        entityKey
      );
      const first = withNewEntity.getBlockMap().first();

      checkForCharacterList(first);
      expect(getEntities(first)).toEqual(
        Immutable.Repeat(entityKey, first.getLength()).toArray()
      );
    }

    it('must apply entity key', () => {
      applyAndCheck('x');
    });

    it('must apply null entity', () => {
      applyAndCheck(null);
    });
  });

  describe('Apply entity across multiple blocks', () => {
    const blockMap = contentState.getBlockMap();
    const first = blockMap.first();
    const last = contentState.getBlockAfter(first.getKey());

    const targetSelection = new SelectionState({
      anchorKey: first.getKey(),
      anchorOffset: 0,
      focusKey: last.getKey(),
      focusOffset: last.getLength(),
    });

    function applyAndCheck(entityKey) {
      const withNewEntity = applyEntityToContentState(
        contentState,
        targetSelection,
        entityKey
      );
      const first = withNewEntity.getBlockMap().first();
      const last = withNewEntity.getBlockAfter(first.getKey());

      checkForCharacterList(first);
      expect(getEntities(first)).toEqual(
        Immutable.Repeat(entityKey, first.getLength()).toArray()
      );
      checkForCharacterList(last);
      expect(getEntities(first)).toEqual(
        Immutable.Repeat(entityKey, last.getLength()).toArray()
      );
    }

    it('must apply entity key', () => {
      applyAndCheck('x');
    });

    it('must apply null entity', () => {
      applyAndCheck(null);
    });
  });
});
