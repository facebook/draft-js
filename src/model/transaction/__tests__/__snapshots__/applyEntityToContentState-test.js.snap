// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply entity key 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must apply entity key accross multiple blocks 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "x",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "x",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "x",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must apply null entity 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must apply null entity key accross multiple blocks 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;
